### ISSUE TITLE
Roslyn doesn't process code analysis attributes on deconstruction methods called in switches

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#39068]
**Version Used**: 

**Steps to Reproduce**:

```cs
#nullable enable
using System.Diagnostics.CodeAnalysis;
class C<T>
{
  void M1(C<string> c)
  {
    var (t1, t2) = c;
    t1.ToString(); // CS8602, correct
    t2.ToString();
  }
  
  void M2(C<string> c)
  {
    switch (c)
    {
      case var (t1, t2):
        t1.ToString(); // no warnings
        t2.ToString();
        break;
    }
  }

  public extern void Deconstruct([MaybeNull] out T t1, out T t2);
}
```

**Expected Behavior**:
`t1.ToString()` has `CS8602 Dereference of a possibly null reference.` warning as it's a variable assigned with `[MaybeNull] out T t1` 

**Actual Behavior**:
No warnings for `M2`

**Comment by jcouv** [comment]
Verified this is fixed by https://github.com/dotnet/roslyn/pull/51001 (merged for 16.10). There is no warning in `M2` now.
Since you're probably going to notice that, the fix is limited to top-level positional patterns ;-) Nested ones (`case (t1, (t2, t3)):`) won't track at the moment.
I'll close the issue. Thanks

**Comment by TessenR** [comment]
@jcouv Sorry for the late reply.
I think there was a misunderstanding regarding the expected behavior here. There _**should**_ be a warning in `M2`. There was no warning when the issue was reported and there's still no warning in
```
Branch main (12 Mar 2021)
Latest commit 8da536d by msftbot[bot]:
Merge pull request #51811 from CyrusNajmabadi/simplifyCompletion

Simplify portions of symbol completion
```

Here's a more specific example exposing the problem. 

**Steps to Reproduce:**

Compile and run the following code:
```cs
#nullable enable
using System.Diagnostics.CodeAnalysis;

class C<T>
{
  public T field;
  public C(T t) => field = t;

  public void Deconstruct([MaybeNull] out T t1,  [MaybeNull] out T t2) => t1 = t2 = default;
}

class Program
{
  public static void Main()
  {
    var c = new C<string>("");
    switch (c)
    {
      case var (t1, t2):
        t1.ToString(); // no warnings; runtime crash
        break;
    }
  }
}
```

**Expected behavior:**
CS8602 warning for `t1.ToString()` since the variable was assigned with a `[MaybeNull] out T t1` parameter which is explicitly annotated as possibly containing a null value

**Actual behavior:**
No warnings reported in the code above. The program crashes at runtime with a `NullReferenceException`

Is it possible to reopen this issue or should I create a new one with this example?

**Comment by jcouv** [comment]
Re-opened. Sorry for the misunderstanding.
